### PR TITLE
Add `AcquisitionFuncParams` for LogEHVI

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Motivation

This PR adds an interface for LogEHVI, a multi-objective acquisition function.

Although this PR passes the CI standalone, this PR should be merged after
- https://github.com/optuna/optuna/pull/6039

### CODE REVIEW

File: optuna/_gp/acqf.py [hunk 1]
```diff
@@ -117,6 +151,72 @@ def from_acqf_params(
...
+        def get_non_dominated_box_bounds(
+            pareto_sols: np.ndarray, ref_point: np.ndarray
+        ) -> tuple[np.ndarray, np.ndarray]:
+            # NOTE(nabenabe): This is a dummy implementation for formatter.
+            return np.ones((5, Y.shape[-1])) * 5.0, np.ones((5, Y.shape[-1])) * 10.0
+
+        def _get_non_dominated_box_bounds() -> tuple[torch.Tensor, torch.Tensor]:
+            loss_vals = -Y  # NOTE(nabenabe): Y is to be maximized, loss_vals is to be minimized.
+            pareto_sols = loss_vals[_is_pareto_front(loss_vals, assume_unique_lexsorted=False)]
+            ref_point = np.max(loss_vals, axis=0) * 1.1
```
Comment (nabenabe0928) [inline optuna/_gp/acqf.py:178]:
```suggestion
            ref_point = np.max(loss_vals, axis=0)
            ref_point = np.maximum(ref_point * 1.1, ref_point * 0.9)
```

**Comment by nabenabe0928** [comment]
@HideakiImamura @kAIto47802 
Could you review this PR?

File: optuna/_gp/acqf.py [hunk 2]
```diff
@@ -148,7 +246,44 @@ def create_acqf_params(
     )
 
 
+def _eval_ehvi(
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py:249]: I've reviewed this function, and it looks sound in principle, though I still need to check it precisely.

File: optuna/_gp/acqf.py [hunk 3]
```diff
@@ -22,6 +24,40 @@
...
+    # https://github.com/pytorch/botorch/blob/v0.13.0/botorch/utils/safe_math.py
+    # https://github.com/pytorch/botorch/blob/v0.13.0/botorch/acquisition/multi_objective/logei.py#L146-L266
+    _EPS = torch.tensor(1e-12, dtype=torch.float64)  # NOTE(nabenabe): grad becomes nan when EPS=0.
+    diff = torch.maximum(
+        _EPS,
+        torch.minimum(Y_post[..., torch.newaxis, :], non_dominated_box_upper_bounds)
+        - non_dominated_box_lower_bounds,
+    )
+    log_hvi_vals = torch.special.logsumexp(diff.log().sum(dim=-1), dim=-1)
+    return -log_n_qmc_samples + torch.special.logsumexp(log_hvi_vals, dim=-1)
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py:58]: This can be simply written as follows:
```suggestion
    return -log_n_qmc_samples + torch.special.logsumexp(diff.log().sum(dim=-1), dim=(-2, -1))
```

File: optuna/_gp/acqf.py [hunk 4]
```diff
@@ -22,6 +24,40 @@
...
+    sobol_engine = torch.quasirandom.SobolEngine(
+        dimension=dim, scramble=True, seed=seed
+    )  # type: ignore
+    # The Sobol sequence in [-1, 1].
+    samples = 2.0 * (sobol_engine.draw(n_samples, dtype=torch.float64) - 0.5)
+    # Inverse transform to standard normal (values to close to 0/1 result in inf values).
+    return torch.erfinv(samples) * float(np.sqrt(2))
+
+
+def logehvi(
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py:38]: I've reviewed this function and almost LGTM.

**Review by kAIto47802 — CHANGES_REQUESTED** [review]

File: optuna/_gp/acqf.py [hunk 5]
```diff
@@ -22,6 +24,41 @@
     torch = _LazyImport("torch")
 
 
+def _sample_from_normal_sobol(dim: int, n_samples: int, seed: int | None) -> torch.Tensor:
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py:27]: I've also reviewed this function, and it appears LGTM.

File: optuna/_gp/acqf.py [hunk 6]
```diff
@@ -117,6 +155,67 @@ def from_acqf_params(
         )
 
 
+@dataclass(frozen=True)
+class MultiObjectiveAcquisitionFunctionParams(AcquisitionFunctionParams):
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py:159]: I've also reviewed this function, and it looks sound in principle.

File: optuna/_gp/acqf.py [hunk 7]
```diff
@@ -117,6 +155,71 @@ def from_acqf_params(
...
+        inverse_squared_lengthscales = np.mean(
+            [
+                acqf_params.kernel_params.inverse_squared_lengthscales.detach().numpy()
+                for acqf_params in acqf_params_for_objectives
+            ],
+            axis=0,
+        )
+        dummy_kernel_params = KernelParamsTensor(
+            # inverse_squared_lengthscales is used in optim_mixed.py.
+            inverse_squared_lengthscales=torch.from_numpy(inverse_squared_lengthscales),
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py:201]: How about keeping it in `torch.Tensor` without temporarily converting it to `np.ndarray`? It would make the implementation simpler.
However, in this case, no copy is performed in both conversions, i.e., both from `torch.Tensor` to `np.ndarray` and from `np.ndarray` to `torch.Tensor`, so it does not incur any overhead. Thus, the current implementation is also acceptable if you prefer.
```suggestion
        inverse_squared_lengthscales = sum(
            acqf_params.kernel_params.inverse_squared_lengthscales.detach().numpy()
            for acqf_params in acqf_params_for_objectives
        ) / len(acqf_params_for_objectives)
        dummy_kernel_params = KernelParamsTensor(
            # inverse_squared_lengthscales is used in optim_mixed.py.
            inverse_squared_lengthscales=inverse_squared_lengthscales,
```

File: optuna/_gp/acqf.py [hunk 8]
```diff
@@ -117,6 +155,71 @@ def from_acqf_params(
...
+    fixed_samples: torch.Tensor
+
+    @classmethod
+    def from_acqf_params(
+        cls,
+        acqf_params_for_objectives: list[AcquisitionFunctionParams],
+        Y: np.ndarray,
+        n_qmc_samples: int,
+        qmc_seed: int | None,
+    ) -> MultiObjectiveAcquisitionFunctionParams:
```
Comment (kAIto47802) [inline optuna/_gp/acqf.py:172]: Although the interface of this dataclass and the naming of this classmethod are open to debate, this section needs refactoring for abstraction anyway. Thus, there is little point in discussing it now, and we can leave it as it is.

**Review by kAIto47802 — CHANGES_REQUESTED** [review]
I have finished reviewing `acqf.py`, leaving some comments. PTAL :octocat:

File: optuna/_gp/acqf.py [hunk 9]
```diff
@@ -117,6 +155,71 @@ def from_acqf_params(
...
+        def _get_non_dominated_box_bounds() -> tuple[torch.Tensor, torch.Tensor]:
+            loss_vals = -Y  # NOTE(nabenabe): Y is to be maximized, loss_vals is to be minimized.
+            pareto_sols = loss_vals[_is_pareto_front(loss_vals, assume_unique_lexsorted=False)]
+            ref_point = np.max(loss_vals, axis=0)
+            ref_point = np.maximum(1.1 * ref_point, 0.9 * ref_point)
```
[the remark this answers is not in the document]
Comment (nabenabe0928) [reply optuna/_gp/acqf.py:177]: Let me modify like this because 1.1/0.9 multiplication is simply the convention (to avoid the hypervolume of zero). 
```suggestion
            ref_point = np.nextafter(np.maximum(1.1 * ref_point, 0.9 * ref_point), np.inf)
```

File: optuna/_gp/acqf.py (hunk 5 above)
Comment (HideakiImamura) [inline optuna/_gp/acqf.py:27]: Uniform sampling based on Sobol sequence and the inverse cumulative distribution function is a good method with fast convergence, especially effective in low-dimensional problems. However, compared to simply sampling normally distributed random variables using methods like `torch.randn`, is this approach really suitable for the current purpose? Are there any references or experimental results that support the use of this method?

File: optuna/_gp/acqf.py [hunk 10]
```diff
@@ -148,7 +251,44 @@ def create_acqf_params(
     )
 
 
+def _eval_ehvi(
+    evhi_acqf_params: MultiObjectiveAcquisitionFunctionParams, x: torch.Tensor
```
Comment (HideakiImamura) [inline optuna/_gp/acqf.py:255]: Typo
```suggestion
    ehvi_acqf_params: MultiObjectiveAcquisitionFunctionParams, x: torch.Tensor
```

File: optuna/_gp/acqf.py [hunk 11]
```diff
@@ -117,6 +155,71 @@ def from_acqf_params(
...
+        # Since all the objectives are equally important, we simply use the mean of
+        # inverse_squared_lengthscales over all the objectives.
+        # cf. https://github.com/optuna/optuna/blob/v4.3.0/optuna/_gp/optim_mixed.py#L200-L209
+        inverse_squared_lengthscales = np.mean(
+            [
+                acqf_params.kernel_params.inverse_squared_lengthscales.detach().numpy()
+                for acqf_params in acqf_params_for_objectives
+            ],
+            axis=0,
+        )
```
Comment (HideakiImamura) [inline optuna/_gp/acqf.py:198]: From what I can see in `_eval_ehvi`, the `kernel_params` in `MultiObjectiveAcquisitionFunctionParams` appears to be a dummy and is not actually used. Therefore, the `inverse_squared_lengthscales` does not need to hold any meaningful values either - would it make sense to simply set it to a dummy tensor, such as `torch.empty(len(params))`?

File: optuna/_gp/acqf.py [hunk 12]
```diff
@@ -117,6 +155,71 @@ def from_acqf_params(
...
+        repr_acqf_params = acqf_params_for_objectives[0]
+        return cls(
+            acqf_type=AcquisitionFunctionType.LOG_EHVI,
+            kernel_params=dummy_kernel_params,
+            X=repr_acqf_params.X,
+            search_space=repr_acqf_params.search_space,
+            cov_Y_Y_inv=np.empty(0),
+            cov_Y_Y_inv_Y=np.empty(0),
+            max_Y=np.nan,  # This is not used anywhere.
+            beta=None,
```
Comment (HideakiImamura) [inline optuna/_gp/acqf.py:214]: All of them are not used anywhere, aren't they? How about adding the excuse comments for all of them.

Also, for `kernel_params`.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I have several comments. PTAL.

File: optuna/_gp/acqf.py (hunk 10 above)
Comment (nabenabe0928) [reply optuna/_gp/acqf.py:255]: Good catch, I did not realize it!

File: optuna/_gp/acqf.py (hunk 11 above)
Comment (nabenabe0928) [reply optuna/_gp/acqf.py:198]: It is a bit tricky because only `inverse_squared_lengthscales` is explicitly accessed externally; please see
- https://github.com/optuna/optuna/blob/v4.3.0/optuna/_gp/optim_mixed.py#L200-L209

So we cannot actually make this tensor a dummy version.

File: optuna/_gp/acqf.py (hunk 5 above)
Comment (HideakiImamura) [reply optuna/_gp/acqf.py:27]:
```
        # NOTE(nabenabe): By using fixed samples from the Sobol sequence, EHVI becomes
        # deterministic, making it possible to optimize the acqf by l-BFGS.
```

This comment helped me understand the necessity of this function. Since the SobolEngine is re-initialized with the same random seed each time, it makes sense that exactly the same points are sampled every time a new instance of MultiObjectiveAcquisitionFunctionParams is created. On the other hand, I'm curious whether it's better to use this approach or to use torch.randn with a newly initialized generator as shown below. What do you think?

```python
g = torch.Generator()
g.manual_seed(seed)
samples = torch.randn(n_samples, dim, dtype=torch.float64, generator=g)
```

File: optuna/_gp/acqf.py (hunk 12 above)
Comment (nabenabe0928) [reply optuna/_gp/acqf.py:214]: I will do so!

File: tests/gp_tests/test_acqf.py [hunk 13]
```diff
@@ -4,15 +4,62 @@
...
+        inverse_squared_lengthscales=torch.tensor([2.0, 3.0], dtype=torch.float64),
+        kernel_scale=torch.tensor(4.0, dtype=torch.float64),
+        noise_var=torch.tensor(0.1, dtype=torch.float64),
+    )
+    search_space = SearchSpace(
+        scale_types=np.full(n_dims, ScaleType.LINEAR),
+        bounds=np.array([[0.0, 1.0] * n_dims]),
+        steps=np.zeros(n_dims),
+    )
+    return kernel_params, search_space, X
```
Comment (HideakiImamura) [inline tests/gp_tests/test_acqf.py:45]: How about defining fixture of them?

File: optuna/_gp/acqf.py (hunk 5 above)
Comment (nabenabe0928) [reply optuna/_gp/acqf.py:27]: The reference is available at Appendix D of [this paper](https://arxiv.org/pdf/2006.05078), but basically, the authors checked four scenarios:

image

File: tests/gp_tests/test_acqf.py (hunk 13 above)
Comment (nabenabe0928) [reply tests/gp_tests/test_acqf.py]: I updated the unit tests accordingly!

File: optuna/_gp/acqf.py (hunk 5 above)
Comment (HideakiImamura) [reply optuna/_gp/acqf.py:27]: I understood. Sounds very good to use QMC fixed samples.

File: optuna/_gp/acqf.py (hunk 11 above)
Comment (HideakiImamura) [reply optuna/_gp/acqf.py:199]: I understood. Thanks for the explanation. BTW, how about using harmonic mean after taking square root? It ensures that the averaged lengthscale is used.

Comment (nabenabe0928) [reply optuna/_gp/acqf.py:199]: Like this?

```python
        inverse_squared_lengthscales_array = np.array(
            [
                acqf_params.kernel_params.inverse_squared_lengthscales.detach().numpy()
                for acqf_params in acqf_params_for_objectives
            ]
        )
        mean_lengthscales = np.mean(1 / np.sqrt(inverse_squared_lengthscales_array))
        dummy_kernel_params = KernelParamsTensor(
            # inverse_squared_lengthscales is used in optim_mixed.py.
            inverse_squared_lengthscales=torch.from_numpy(1.0 / mean_lengthscales**2),
            # These parameters will not be used anywhere.
            kernel_scale=torch.empty(0),
            noise_var=torch.empty(0),
        )
```

Comment (nabenabe0928) [reply optuna/_gp/acqf.py]: I applied your suggestion with the co-authorship:)

**Review by HideakiImamura — APPROVED** [review]
Thanks for the update! LGTM!

File: optuna/_gp/acqf.py (hunk 7 above)
Comment (nabenabe0928) [reply optuna/_gp/acqf.py]: As `np.mean` is more intuitive compared to `sum(...) / len(...)` and `mean(stack(...))`, I decided to stick to the current implementation, but thank you for the suggestion:)

File: optuna/_gp/acqf.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kAIto47802) [reply optuna/_gp/acqf.py]: Thank you for your reply. I didn't know about the convention.

**Review by kAIto47802 — APPROVED** [review]
Thank you for the update! LGTM
